File: 0567. Permutation in String.cpp
class Solution
{
public:
    bool checkInclusion(string s1, string s2)
    {
        if (s1.size() > s2.size())
        {
            return false;
        }
        
        vector<int> m1(26, 0);
        vector<int> m2(26, 0);
        
        int ind = 0;
        
        for (; ind < s1.size(); ++ind)
        {
            ++m1[s1[ind] - 'a'];
            ++m2[s2[ind] - 'a'];
        }
        
        if (m1 == m2)
        {
            return true;
        }
        
        for (; ind < s2.size(); ++ind)
        {
            ++m2[s2[ind] - 'a'];
            --m2[s2[ind - s1.size()] - 'a'];
            
            if (m1 == m2)
            {
                return true;
            }
        }
        
        return false;
    }
};

File: 0567. Permutation in String.cs
public class Solution
{
    public bool CheckInclusion(string s1, string s2)
    {
        var frequencies = new int[26];

        foreach (var c in s1)
        {
            ++frequencies[c - 'a'];
        }

        var l = 0;
        var r = 0;
        var currentFrequencies = new int[26];

        while (r < s2.Length)
        {
            var current = s2[r] - 'a';
            ++currentFrequencies[current];

            while (currentFrequencies[current] > frequencies[current])
            {
                --currentFrequencies[s2[l] - 'a'];
                ++l;
            }

            if (frequencies.SequenceEqual(currentFrequencies))
            {
                return true;
            }

            ++r;
        }

        return false;
    }
}